haskell/asterius/defs.bzl

```diff
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
     },
     doc = "Toolchain for asterius tools that are not part of the regular haskell toolchain",
 )
+
+# The ahc_dist rule generates an archive containing javascript files
+# from a haskell binary build with the asterius toolchain.  We ensure
+# that this toolchain is selected using the following transition to
+# select the asterius platform
+def _asterius_transition(settings, attr):
+    return {"//command_line_option:platforms": "@rules_haskell//haskell/asterius:asterius_platform"}
+
+asterius_transition = transition(
+    implementation = _asterius_transition,
+    inputs = [],
+    outputs = ["//command_line_option:platforms"],
+)
+
+AhcDistInfo = provider(
+    "Info about the output files of ahc_dist.",
+    fields = {
+        "target": "one of node or desktop",
+        "webpack_config": "The webpack_config file that may be used by the asterius_webpack rule ",
+        "files": "other files from the archive",
+    },
+)
+
+# runtime modules generated by ahc_dist
+javascript_runtime_modules = [
+    "rts.autoapply.mjs",
+    "rts.closuretypes.mjs",
+    "rts.constants.mjs",
+    "rts.eventlog.mjs",
+    "rts.exception.mjs",
+    "rts.exports.mjs",
+    "rts.float.mjs",
+    "rts.fs.mjs",
+    "rts.funtypes.mjs",
+    "rts.gc.mjs",
+    "rts.heapalloc.mjs",
+    "rts.integer.mjs",
+    "rts.jsval.mjs",
+    "rts.memory.mjs",
+    "rts.memorytrap.mjs",
+    "rts.messages.mjs",
+    "rts.mjs",
+    "rts.modulify.mjs",
+    "rts.reentrancy.mjs",
+    "rts.scheduler.mjs",
+    "rts.setimmediate.mjs",
+    "rts.stablename.mjs",
+    "rts.stableptr.mjs",
+    "rts.staticptr.mjs",
+    "rts.symtable.mjs",
+    "rts.time.mjs",
+    "rts.tracing.mjs",
+    "rts.unicode.mjs",
+    "rts.wasi.mjs",
+    "default.mjs",
+]
+
+# Label of the template file to use for the webpack config.
+_TEMPLATE = "//haskell/asterius:asterius_webpack_config.js.tpl"
+
+def _ahc_dist_impl(ctx):
+    asterius_toolchain = ctx.toolchains["@rules_haskell//haskell/asterius:toolchain_type"]
+    posix_toolchain = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
+    nodejs_toolchain = ctx.toolchains["@build_bazel_rules_nodejs//toolchains/node:toolchain_type"]
+    node_toolfiles = nodejs_toolchain.nodeinfo.tool_files
+
+    all_output_files = [ctx.attr.main_output]
+    for m in javascript_runtime_modules:
+        all_output_files.append(m)
+
+    (output_prefix, _) = paths.split_extension(ctx.attr.main_output)
+    output_tar_file = ctx.actions.declare_file("{}.tar".format(ctx.label.name))
+    for ext in [".wasm", ".wasm.mjs", ".req.mjs"]:
+        all_output_files.append(output_prefix + ext)
+
+    # ctx.file.dep was generated in the asterius platform configuration,
+    # and we want to generate js files in the current configuration.
+    # So we will copy it to the folder corresponding to the current platform.
+    file_copy_path = paths.join(output_tar_file.dirname, ctx.file.dep.basename)
```

Comment:
This should be fine now that we output files in a subfolder.
At the moment, it is not possible to set `--platforms @rules_haskell//haskell/asterius:asterius_platform` globally when compiling a `ahc_dist` target because it depends on the `rules_nodejs` toolchain that sets a target constraint (and maybe should not). But only the dependencies of this target need to be built for the asterius_platform.